features/filesystem/bd/HeapBlockDevice.h

```diff
@@ -0,0 +1,132 @@
+#define MBED_MEM_BLOCK_DEVICE_H
+
+#include "BlockDevice.h"
+#include "mbed.h"
+
+
+/** Lazily allocated heap-backed block device
+ *
+ * Useful for simulating a block device and tests
+ *
+ * @code
+ * #include "mbed.h"
+ * #include "HeapBlockDevice.h"
+ *
+ * HeapBlockDevice bd(2048, 512); // 2048 bytes with a block size of 512 bytes
+ * uint8_t block[512] = "Hello World!\n";
+ *
+ * int main() {
+ *     bd.init();
+ *     bd.write(block, 0);
+ *     bd.read(block, 0);
+ *     printf("%s", block);
+ *     bd.deinit();
+ * }
+ */
+class HeapBlockDevice : public BlockDevice {
+public:
+
+    /** Lifetime of the memory block device
+     */
+    HeapBlockDevice(bd_size_t size, bd_size_t block=512);
+    HeapBlockDevice(bd_size_t size, bd_size_t read, bd_size_t program, bd_size_t erase);
+    virtual ~HeapBlockDevice();
+
+    /** Initialize a block device
+     *
+     *  @return         0 on success or a negative error code on failure
+     */
+    virtual bd_error_t init();
+
+    /** Deinitialize a block device
+     *
+     *  @return         0 on success or a negative error code on failure
+     */
+    virtual bd_error_t deinit();
+
+    /** Read blocks from a block device
+     *
+     *  @param buffer   Buffer to write blocks to
+     *  @param addr     Address of block to begin reading from
+     *  @param size     Size to read in bytes, must be a multiple of read block size
+     *  @return         0 on success, negative error code on failure
+     */
+    virtual bd_error_t read(void *buffer, bd_addr_t addr, bd_size_t size);
+
+    /** Program blocks to a block device
+     *
+     *  The blocks must have been erased prior to being programmed
+     *
+     *  @param buffer   Buffer of data to write to blocks
+     *  @param addr     Address of block to begin writing to
+     *  @param size     Size to write in bytes, must be a multiple of program block size
+     *  @return         0 on success, negative error code on failure
+     */
+    virtual bd_error_t program(const void *buffer, bd_addr_t addr, bd_size_t size);
+
+    /** Erase blocks on a block device
+     *
+     *  The state of an erased block is undefined until it has been programmed
+     *
+     *  @param addr     Address of block to begin erasing
+     *  @param size     Size to erase in bytes, must be a multiple of erase block size
+     *  @return         0 on success, negative error code on failure
+     */
+    virtual bd_error_t erase(bd_addr_t addr, bd_size_t size);
+
+    /** Get the size of a readable block
+     *
+     *  @return         Size of a readable block in bytes
+     */
+    virtual bd_size_t get_read_size();
+
+    /** Get the size of a programable block
+     *
+     *  @return         Size of a programable block in bytes
+     */
+    virtual bd_size_t get_program_size();
+
+    /** Get the size of a eraseable block
+     *
+     *  @return         Size of a eraseable block in bytes
+     */
+    virtual bd_size_t get_erase_size();
```

Comment:
They could be, I originally did not since it may require implementors learn to use mutable for cases where non-const SPI transaction may occur.

If we're pushing for const (and more adoption of const+mutable where simple hardware transaction are required) I don't see a reason to not slap const on these guys.